AnnoDesigner.Core/DataStructures/QuadTree.cs

```diff
@@ -357,154 +359,184 @@ public void ReIndex(Func<T, Rect> boundsSelector)
-            root.Insert(item, bounds);
+            var bounds = item.Bounds;
+            EnsureBounds(bounds);
+
+            root.Add(item, bounds);
         }
 
         /// <summary>
         /// Remove a <typeparamref name="T"/> item from the QuadTree.
         /// </summary>
-        /// <param name="item"></param>
-        /// <param name="bounds"></param>
-        public void Remove(T item, Rect bounds)
+        public bool Remove(T item)
         {
-            var quadrant = root.GetContainingQuadrant(bounds);
-            quadrant.Remove((item, bounds));
+            return root.Remove(item);
         }
 
         /// <summary>
-        /// Retrieves a sequence of items that intersect the given bounds.
+        /// Removes all items from the QuadTree.
         /// </summary>
-        /// <param name="bounds">The bounds to check</param>
-        /// <returns></returns>
-        public IEnumerable<T> GetItemsIntersecting(Rect bounds)
+        public void Clear()
         {
-            var items = new List<T>(previousCount);
-            root.GetItemsIntersecting(items, bounds);
-            previousCount = items.Count;
-            return items;
+            foreach (var item in this.ToList())
+            {
+                Remove(item);
+            }
         }
 
+        public bool Contains(T item) => root.Contains(item);
+
+        public void CopyTo(T[] array, int arrayIndex) => root.CopyTo(array, arrayIndex);
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        public IEnumerator<T> GetEnumerator() => root.GetEnumerator();
+
         /// <summary>
-        /// Retrieves all the items from the <see cref="QuadTree{T}"/>
+        /// Ensures specified bound is in Extent of this QuadTree.
+        /// Extent is inflated in direction depending on in which direction <paramref name="bounds"/> is
+        /// 
+        /// TL|TR|TR
+        /// --------
+        /// TL|  |BR
+        /// --------
+        /// BL|BL|BR
+        /// 
+        /// (middle of the grid represents old extent)
+        /// 
         /// </summary>
-        /// <returns></returns>
-        public IEnumerable<T> All()
+        public void EnsureBounds(Rect bounds)
         {
-            return root.All();
-        }
-
-        public IEnumerable<T> FindAll(Predicate<T> match)
-        {
-            if (match is null)
+            while (!Extent.Contains(bounds))
             {
-                throw new ArgumentNullException($"{nameof(match)}");
-            }
-            var list = new List<T>();
-            foreach (var item in this)
-            {
-                if (match(item))
+                var newExtent = Extent;
+                newExtent.Union(bounds);
+                var top = Extent.Top - newExtent.Top > 0;
+                var bottom = newExtent.Bottom - Extent.Bottom > 0;
+                var left = Extent.Left - newExtent.Left > 0;
+                var right = Extent.Right - newExtent.Right > 0;
```

Comment:
This piece of code looks a little confusing to me - could maybe do with a refactor or a comment?

Reply:
How about both? And a bugfix for free!